### ISSUE TITLE
Unable to build kernel module with `--compile` flag on AKS (Ubuntu 18.04; Kubernetes 1.24)

### ISSUE THREAD

**Issue by tspearconquest** [falcosecurity/falco#2467]
**Summary**
I've recently tried to switch a test cluster from the ebpf probe to the kernel module and have it compile the module with the driver loader as an init container using the `--compile` flag. This was done to provide thorough testing of a slimmed down custom falco-driver-loader image for our internal use without unnecessary extras. We wanted to be able to build the module or ebpf probe from source so that we could have a driver-loader-ebpf image and a driver-loader-dkms image with only the minimal necessary requirements to build the module/probe and load it.

**Describe the bug**
The kernel module build errors out for me during container startup and I am not clear on what's wrong.

**How to reproduce it**
The module compilation fails

**Expected behaviour**
Module compilation should succeed and the module should load into memory

a**Screenshots**

**Environment**

- Falco version: 0.34.1

- System info: 

- Cloud provider or hardware configuration: Azure AKS 1.24
- OS: Ubuntu 18.04

- Kernel: 5.4.x

- Installation method: Kubernetes

**Additional context**

```
logs -n falco -c driver-loader po/falco-dktpt                                                                        3s ⎈ armedatk-centralus-dev-01
* Setting up /usr/src links from host
* Running falco-driver-loader for: falco version=0.34.1, driver version=4.0.0+driver, arch=x86_64, kernel release=5.4.0-1103-azure, kernel version=109
* Running falco-driver-loader with: driver=module, compile=yes, download=no

================ Cleaning phase ================

* 1. Check if kernel module 'falco' is still loaded:
- OK! There is no 'falco' module loaded.

* 2. Check all versions of kernel module 'falco' in dkms:
- OK! There are no 'falco' module versions in dkms.

[SUCCESS] Cleaning phase correctly terminated.

================ Cleaning phase ================

* Looking for a falco module locally (kernel 5.4.0-1103-azure)
* Filename 'falco_ubuntu-azure_5.4.0-1103-azure_109.ko' is composed of:
 - driver name: falco
 - target identifier: ubuntu-azure
 - kernel release: 5.4.0-1103-azure
 - kernel version: 109
install: /usr/lib/gcc/x86_64-linux-gnu/5/
* Trying to dkms install falco module with GCC /usr/bin/gcc
DIRECTIVE: MAKE="'/tmp/falco-dkms-make'"

Creating symlink /var/lib/dkms/falco/4.0.0+driver/source ->
                 /usr/src/falco-4.0.0+driver

DKMS: add completed.

Kernel preparation unnecessary for this kernel.  Skipping...

Building module:
cleaning build area...
'/tmp/falco-dkms-make'...(bad exit status: 2)
* Running dkms build failed, dumping /var/lib/dkms/falco/4.0.0+driver/build/make.log (with GCC /usr/bin/gcc)
DKMS make.log for falco-4.0.0+driver for kernel 5.4.0-1103-azure (x86_64)
Mon Apr  3 19:54:46 UTC 2023
'/tmp/falco-dkms-make' -C /lib/modules/5.4.0-1103-azure/build M=/var/lib/dkms/falco/4.0.0+driver/build modules
make[1]: Entering directory '/host/usr/src/linux-headers-5.4.0-1103-azure'
  CC [M]  /var/lib/dkms/falco/4.0.0+driver/build/main.o
In file included from <command-line>:0:0:
./arch/x86/include/asm/page_64.h: In function 'clear_page':
././include/linux/compiler_types.h:214:24: error: expected '(' before '__inline'
 #define asm_inline asm __inline
                        ^
./arch/x86/include/asm/alternative.h:256:2: note: in expansion of macro 'asm_inline'
  asm_inline volatile (ALTERNATIVE_2("call %P[old]", "call %P[new1]", feature1,\
  ^
./arch/x86/include/asm/page_64.h:49:2: note: in expansion of macro 'alternative_call_2'
  alternative_call_2(clear_page_orig,
  ^
In file included from ./arch/x86/include/asm/barrier.h:5:0,
                 from ./include/linux/compiler.h:325,
                 from ./include/linux/export.h:42,
                 from ./include/linux/linkage.h:7,
                 from ./include/linux/kernel.h:8,
                 from ./include/linux/list.h:9,
                 from ./include/linux/smp.h:12,
                 from ./include/linux/tracepoint.h:15,
                 from ./include/trace/syscall.h:5,
                 from /var/lib/dkms/falco/4.0.0+driver/build/main.c:13:
./arch/x86/include/asm/alternative.h:128:2: error: expected identifier or '(' before string constant
  "# ALT: oldinstr2\n"         \
  ^
./arch/x86/include/asm/alternative.h:168:2: note: in expansion of macro 'OLDINSTR_2'
  OLDINSTR_2(oldinstr, 1, 2)     \
  ^
./arch/x86/include/asm/alternative.h:256:23: note: in expansion of macro 'ALTERNATIVE_2'
  asm_inline volatile (ALTERNATIVE_2("call %P[old]", "call %P[new1]", feature1,\
                       ^
./arch/x86/include/asm/page_64.h:49:2: note: in expansion of macro 'alternative_call_2'
  alternative_call_2(clear_page_orig,
  ^
In file included from <command-line>:0:0:
././include/linux/compiler_types.h:214:24: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 #define asm_inline asm __inline
                        ^
./arch/x86/include/asm/alternative.h:256:2: note: in expansion of macro 'asm_inline'
  asm_inline volatile (ALTERNATIVE_2("call %P[old]", "call %P[new1]", feature1,\
  ^
./arch/x86/include/asm/page_64.h:49:2: note: in expansion of macro 'alternative_call_2'
  alternative_call_2(clear_page_orig,
  ^
```

These logs repeat numerous times with different files calling into the same headers

**Comment by Andreagit97** [comment]
Uhm that's interesting. To install the kernel module locally you need 3 main things:
* dkms
* kernel headers
* a recent version of GCC (it really depends on the machine)
 
Do you have `kernel-headers` installed? if yes, I would suggest you try a more recent `gcc` version

**Comment by tspearconquest** [comment]
It looks like the node's kernel headers are mounted in the container. From the above log:
`make[1]: Entering directory '/host/usr/src/linux-headers-5.4.0-1103-azure'`

**Comment by Andreagit97** [comment]
Yeah that's true. It seems like a problem with the GCC version you probably need to use a more recent version like `gcc-11` or `gcc-12`. In `driverkit` we use these 2 versions to build drivers `5.X` https://github.com/falcosecurity/driverkit/blob/master/pkg/driverbuilder/builder/builders.go#L123

**Comment by FedeDP** [comment]
This is something that will be addressed by https://github.com/falcosecurity/falco/issues/2489.

**Comment by epcim** [comment]
The issue really looks like wrong headers are installed.


As the issue is recent I would mention few things, not directly related (but worth to consider if anybody touching codebase). Sure I can open new issue for them - as feature request.

Speaking about  `/usr/bin/falco-driver-loader --compile`

It would be great if falco-driver-loader script/function would accept ENV variable to override KERNELDIR Makefile variable and used it for the line
```
* Trying to dkms install falco module with GCC /usr/bin/gcc
DIRECTIVE: MAKE="'/tmp/falco-dkms-make'"
```

so one would be able from POD env modify the kernel path  used for compile. Not sure how on Azure, but on AWS it's frequent in my env that packages on host looks like this (latest):

```
[root@ip-172-16-184-249 ~]# yum list installed -v |grep 'kernel' | grep '5.4'
kernel.x86_64                         5.4.238-148.346.amzn2          @amzn2extra-kernel-5.4
kernel-devel.x86_64                   5.4.238-148.347.amzn2          @amzn2extra-kernel-5.4
kernel-headers.x86_64                 5.4.238-148.347.amzn2          @amzn2extra-kernel-5.4
```


^^ I am runnig kernel  `5.4.238-148.346` but there are not headers for it and the DKMS build fails to build it properly because this link fails. On AWS I am either not able to install `346` devel/headers as packages are not available.

The makefile poits KERNELDIR to (as `uname -a` is used)
```
root@falco2-xx6pv:/var/lib/dkms/falco/4.0.0+driver# ls -l /lib/modules/5.4.238-148.346.amzn2.x86_64/build
lrwxrwxrwx 1 root root 45 Apr  6 01:42 /lib/modules/5.4.238-148.346.amzn2.x86_64/build -> /usr/src/kernels/5.4.238-148.346.amzn2.x86_64

```

while against 347 it would work ie: `root@falco2-xx6pv:/var/lib/dkms/falco/4.0.0+driver/source# make -C /usr/src/kernels/5.4.238-148.347.amzn2.x86_64  M=/usr/src/falco-4.0.0+driver  modules`

**Comment by epcim** [comment]
RC on AWS side not. We have just upgraded all pkgs to latest. 

It might be some release “skew” on AWS side. 

Or 346 is some minor versioning, so it might be even ok. Falco is wrong on using kernel dir through wrong link. 

People are freq. complaining about failing builds. Strange. 

Compile works actually. 

this is sort of fix for makefile falback I used. 
```

KERNELDIR	?= /lib/modules/$(shell uname -r)/build
# WORKAROUND (latest available kernel installed)
KERNELALT 		?= /host/usr/src/kernels/$(shell ls /host/usr/src/kernels|sort|tail -n1)
ifeq ($(wildcard $(KERNELDIR)/*),)
	KERNELDIR   := $(KERNELALT)
endif
```

**Comment by epcim** [comment]
Btw is there a way to prefer build over download. Seems to me download is hardcoded as first if both flags are in place.

**Comment by FedeDP** [comment]
You can enforce the usage of the compile-only by passing `--compile` flag to falco-driver-loader.

Btw for any feature/improvement request, please feel free to open an issue. `falco-driver-loader` script will, sooner or later, be integrated in [falcoctl](https://github.com/falcosecurity/falcoctl/), becoming a full go program instead of a bash script.

**Comment by epcim** [comment]
> You can enforce the usage of the compile-only by passing `--compile` flag to falco-driver-loader.
> 
> Btw for any feature/improvement request, please feel free to open an issue. `falco-driver-loader` script will, sooner or later, be integrated in [falcoctl](https://github.com/falcosecurity/falcoctl/), becoming a full go program instead of a bash script.

as it will change, less prone to push PR

you can pass only --compile, but that is not soulution if you want "download" ad fall-back if compile fail..

**Comment by epcim** [comment]
> RC on AWS side not. We have just upgraded all pkgs to latest.
> 
> It might be some release “skew” on AWS side.
> 
> Or 346 is some minor versioning, so it might be even ok. Falco is wrong on using kernel dir through wrong link.
> 
> People are freq. complaining about failing builds. Strange.
> 
> Compile works actually.
> 
> this is sort of fix for makefile falback I used.
> 
> ```
> 
> KERNELDIR	?= /lib/modules/$(shell uname -r)/build
> # WORKAROUND (latest available kernel installed)
> KERNELALT 		?= /host/usr/src/kernels/$(shell ls /host/usr/src/kernels|sort|tail -n1)
> ifeq ($(wildcard $(KERNELDIR)/*),)
> 	KERNELDIR   := $(KERNELALT)
> endif
> ```

that only works in /usr/src/falco-4.0.0+driver

dkms will fail anyway:
```
root@falco-wq96b:/usr/src/falco-4.0.0+driver# dkms install --directive="MAKE='/tmp/falco-dkms-make KERNELDIR=hovno'" -m "${DRIVER_NAME}" -v "${DRIVER_VERSION}" -k "${KERNEL_RELEASE}"
DIRECTIVE: MAKE="'/tmp/falco-dkms-make"
Error! echo
Your kernel headers for kernel 5.4.238-148.346.amzn2.x86_64 cannot be found at
/lib/modules/5.4.238-148.346.amzn2.x86_64/build or /lib/modules/5.4.238-148.346.amzn2.x86_64/source.
```


the issue, cant be fixed in the container - as long as the proper sources are missing - even build itself would work

**Comment by Andreagit97** [comment]
This is something that should be addressed by https://github.com/falcosecurity/falco/issues/2489.

**Comment by FedeDP** [comment]
Yep, this should be fixed using latest Falco 0.36.0 images!

**Comment by Andreagit97** [comment]
[Falco 0.36](https://github.com/falcosecurity/falco/releases/tag/0.36.0) should solve this issue, I will close it. Feel free to reopen if this is not solved :)!
